### DISCUSSION TITLE
[Question][Hw2] ill-typed operations

### DISCUSSION THREAD

**Discussion by KihongHeo** [prosyslab-classroom/cs524-program-analysis#413] [Q&A]
Answer to Seungmin's last question:

You have to raise exception `Invalid_argument` when operations are ill-typed, such as `1 + True`.

**Comment by h2oche** [comment]
I got one more question. What is the semantics of arithmetic operations of `Bool`? For example `True + True` or `True < False`.

**Comment by KihongHeo** [comment]
Also `Invalid_argument`.
In short, you can use `Invalid_argument` for all undefined behavior.
Exception: `Division_by_zero`.

See README.md.

**Comment by h2oche** [comment]
Thanks for clarification!

**Comment by goodtaeeun** [comment]
How can we raise ```Invalid_argument``` for ill-typed operations?

When I try to add boolean value with integer with the following code,

```
int a = 9;
int b = 4;
print( (a<b) + a);
``` 

I get the following llvm IR

```
entry:
  %cmp = icmp slt i32 9, 4
  %conv = zext i1 %cmp to i32
  %add = add nsw i32 %conv, 9
  call void @print(i32 %add)
  ret i32 0
```

Since ```zext``` is not handled by our code, all I get is ```Unsupported``` exception.

**Comment by KihongHeo** [comment]
You can ignore all programs that have unsupported instructions.

**Comment by goodtaeeun** [comment]
Thank you!
